### PR TITLE
[Target 4.10.1] Support Content Security Policy

### LINKED ISSUE #3393: Adding nonce to style to support Content Security Policy
## Feature Request

To enable Content Security Policy on the hosting page, Web Chat will need to be updated with nonce in the styles, i.e. changing the prefix for glamor.

## Additional Context

https://portal.microsofticm.com/imp/v3/incidents/details/199836319/home

[Enhancement]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #3393.

### Breaking changes

-  To support Content Security Policy, [`glamor`](https://npmjs.com/package/glamor) is being replaced by [`create-emotion`](https://npmjs.com/package/create-emotion). The CSS hash and rule name is being prefixed with `webchat--css` with a random value.

### Changed

-  Support Content Security Policy, in PR [#3443](https://github.com/microsoft/BotFramework-WebChat/pull/3443) by [@compulim](https://github.com/compulim)
   -  Moved from [`glamor@2.20.40`](https://npmjs.com/package/glamor) to [`create-emotion@10.0.27`](https://npmjs.com/package/create-emotion)
   -  Inlined assets are now using `blob:` scheme, instead of `data:` scheme
   -  Detect Web Worker support by loading a dummy Web Worker, instead of checking `window.MessagePort` and `window.Worker`
   -  Data URI used in image of attachments will be converted to `blob:`
   -  Bumped dependencies
      -  [`react-film@3.0.0`](https://npmjs.com/package/react-film)
      -  [`react-scroll-to-bottom@4.0.0`](https://npmjs.com/package/react-scroll-to-bottom)

### Samples

-  Added Content Security Policy sample, by [@compulim](https://github.com/compulim), in PR [#3443](https://github.com/microsoft/BotFramework-WebChat/pull/3443)

## Description

This feature enables Web Chat to be embedded on a page with Content Security Policy enabled.

## Design

Please refer to [`CONTENT_SECURITY_POLICY.md`](https://github.com/microsoft/BotFramework-WebChat/tree/master/docs/CONTENT_SECURITY_POLICY.md) for design.

## Specific Changes

- Defined a baseline CSP
- Added a new prop `nonce`
   - Currently, this is for injecting `<style>` element, but not `` and other media tags (see/vote #3445)
- Moved from [`glamor`](https://npmjs.com/package/glamor) to [`create-emotion`](https://npmjs.com/package/create-emotion)
   - Updated CSS prefix from `css-*` to `webchat--css-xxxxx-*`
   - The CSS prefix will now have a randomized value to support multiple nonces
- Inlined assets are now using `blob:` scheme, instead of `data:` scheme
   - CSP directive `img-src data:` [is insecure](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src)
- Image attachments using data URI will be parsed and converted to URL with scheme of `blob:`
- Web Worker detection is now done by loading a dummy Web Worker, instead of checking `window.Worker`
- Bumped dependencies to support CSP
   - [`react-film@3.0.0`](https://npmjs.com/package/react-film)
   - [`react-scroll-to-bottom@4.0.0`](https://npmjs.com/package/react-scroll-to-bottom)
- Added new internal hook `useStyleToEmotionObject` to convert from [object style](https://emotion.sh/docs/object-styles) to `emotion` object
   - This is internal as we do not want to take [`emotion`](https://emotion.sh) as part of our long-term dependency, this is also true for `glamor` and most of our dependencies
- Added new internal hook `useNonce` for `nonce` prop

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] Browser and platform compatibilities reviewed
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] Internationalization reviewed (strings, unit formatting)
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: docs/CONTENT_SECURITY_POLICY.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Content Security Policy
+
+Starting from 4.10.1, [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) is supported.
+
+> In this article, we are using a nonce of `a1b2c3d`. In production system, it should be a random and unguessable value that would change every time a policy is applied.
```
Comment (corinagum) [inline docs/CONTENT_SECURITY_POLICY.md:5]:
```suggestion
> In this article, we are using a nonce of `a1b2c3d`. In a production system, it should be a random and unguessable value that changes every time a policy is applied.
```

File: docs/CONTENT_SECURITY_POLICY.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+<!-- prettier-ignore-start -->
+```html
+<meta
+  http-equiv="Content-Security-Policy"
+  content="default-src 'none'; base-uri 'none'; connect-src blob: https://directline.botframework.com wss://directline.botframework.com; img-src blob:; script-src 'strict-dynamic'; style-src 'nonce-a1b2c3d'"
+/>
+```
+<!-- prettier-ignore-end -->
+
+> Additional source for `script-src` will be needed depends on nonce or `'self'` source is used to load Web Chat.
```
Comment (corinagum) [inline docs/CONTENT_SECURITY_POLICY.md:24]:
```suggestion
> Additional source for `script-src` will be needed depending on whether nonce or `'self'` source is used to load Web Chat.
```

File: packages/component/src/Activity/SayAlt.js [hunk 3]
```diff
@@ -1,14 +1,19 @@
-import { css } from 'glamor';
 import PropTypes from 'prop-types';
 import React from 'react';
 
+import useStyleToEmotionObject from '../hooks/internal/useStyleToEmotionObject';
+
 // TODO: [P3] Although this is for development purpose, prettify it
```
Comment (corinagum) [inline packages/component/src/Activity/SayAlt.js:6]: is this still needed?

File: packages/component/src/Composer.js [hunk 4]
```diff
@@ -319,9 +323,23 @@ const Composer = ({
     transcriptFocusRef
   ]);
 
+  const styleToEmotionObject = useMemo(() => {
+    // Emotion don't hash with nonce. We need to provide the pooling mechanism.
```
Comment (corinagum) [inline packages/component/src/Composer.js:327]:
```suggestion
    // Emotion doesn't hash with nonce. We need to provide the pooling mechanism.
```

**Review by corinagum — APPROVED** [review]
approved pending a couple comments

File: packages/component/src/Activity/SayAlt.js (hunk 3 above)
Comment (compulim) [reply packages/component/src/Activity/SayAlt.js:6]: Yes, we still need this.

**Pull request merged**